tests/test_BT_piecewise.py

```diff
@@ -0,0 +1,560 @@
+            j=i,
+        )
+        piecewise_model.add_piecewise_param(
+            "A1", "ls", piecewise_model.A1.value + i * 1e-5, i
+        )
+        piecewise_model.add_piecewise_param(
+            "T0", "d", piecewise_model.T0.value + i * 1e-5, i
+        )
+    return piecewise_model
+
+
+def test_add_piecewise_parameter(build_model):
+    # test: see if the model can be reproduced after piecewise parameters have been added,
+    # checks by comparing parameter keys in both the old and new file. Should have the number of matches = number of parameters
+    m_piecewise = deepcopy(build_model)
+    n = 10
+    m_piecewise.remove_range(0)
+    for i in range(0, n):
+        m_piecewise.add_group_range(
+            m_piecewise.START.value + 501 * i, m_piecewise.START.value + 1000 * i, j=i
+        )
+        m_piecewise.add_piecewise_param("A1", "ls", m_piecewise.A1.value + 1, i)
+        m_piecewise.add_piecewise_param("T0", "d", m_piecewise.T0.value, i)
+    m3 = get_model(StringIO(m_piecewise.as_parfile()))
+    param_dict = m_piecewise.get_params_dict(which="all")
+    copy_param_dict = m3.get_params_dict(which="all")
+    number_of_keys = 0
+    comparison = 0
+    for key, value in param_dict.items():
+        number_of_keys = number_of_keys + 1  # iterates up to total number of keys
+        for copy_key, copy_value in param_dict.items():
+            if key == copy_key:  # search both pars for identical keys
+                if value.quantity == copy_value.quantity:
+                    comparison = (
+                        comparison + 1
+                    )  # iterates up to all matched keys, which should be all, hence total number of keys
+    assert comparison == number_of_keys  # assert theyre the same length
+
+
+def test_get_number_of_groups(build_piecewise_model_with_two_pieces):
+    # test to make sure number of groups matches with number of added piecewise parameters
+    m_piecewise = build_piecewise_model_with_two_pieces
+    number_of_groups = get_number_of_groups(m_piecewise)
+    assert number_of_groups == 2
+
+
+def test_group_assignment_toas_unambiguously_within_group(
+    build_piecewise_model_with_two_pieces, make_toas_to_go_with_two_piece_model
+):
+    # test to see if the group, for one toa per group for 5 groups, that the BT_piecewise.print_toas_in_group functions as intended.
+    # operates by sorting the toas by MJD compared against a groups upper/lower edge.
+    # operates with np.searchsorted so for 1 toa per group, each toa should be uniquely indexed after/before the lower/upper edge
+    index = get_toa_group_indexes(
+        build_piecewise_model_with_two_pieces, make_toas_to_go_with_two_piece_model
+    )
+    print(index)
+    should_be_ten_toas_in_each_group = [
+        np.unique(index, return_counts=True)[1][0],
+        np.unique(index, return_counts=True)[1][1],
+    ]
+    expected_toas_in_each_group = [10, 10]
+    is_there_ten_toas_per_group = np.array_equal(
```

Comment:
it's better to use `numpy.testing.assert_array_equal`, it will tell you more about the mismatch